rust/connlib/snownet/src/stun_binding.rs

```diff
@@ -203,58 +250,119 @@ mod tests {
+        stun_binding.handle_timeout(now);
         assert!(stun_binding.poll_transmit().is_some());
-        assert!(stun_binding.poll_transmit().is_none());
 
         assert_eq!(
             stun_binding.poll_timeout().unwrap(),
-            start + Duration::from_secs(5)
+            now + Duration::from_secs(5),
+            "backoff should be back to 5 seconds timeout"
         );
+    }
 
-        // Nothing after 1 second ..
-        stun_binding.handle_timeout(start + Duration::from_secs(1));
-        assert!(stun_binding.poll_transmit().is_none());
+    #[test]
+    fn retries_requests_using_backoff_and_gives_up_eventually() {
+        let start = Instant::now();
+        let mut stun_binding = StunBinding::new(SERVER1, start);
 
-        // Nothing after 2 seconds ..
-        stun_binding.handle_timeout(start + Duration::from_secs(2));
-        assert!(stun_binding.poll_transmit().is_none());
+        fn secs(secs: f64) -> Duration {
+            Duration::from_micros((secs * 1_000_000.0) as u64)
+        }
 
-        stun_binding.handle_timeout(start + Duration::from_secs(5));
-        assert!(stun_binding.poll_transmit().is_some());
-        assert!(stun_binding.poll_transmit().is_none());
+        // The backoff strategy is to increment the previous interval by 1.5
+        let mut expected_backoffs = VecDeque::from([
+            start + secs(5.0),
+            start + secs(5.0 + 7.5),
+            start + secs(5.0 + 7.5 + 11.25),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 1.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 2.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 3.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 4.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 5.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 6.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 7.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 8.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 9.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 10.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 11.0),
+            start + secs(5.0 + 7.5 + 11.25 + 16.875 + 25.3125 + 37.96875 + 56.953125 + 60.0 * 12.0),
+        ]);
```

Comment:
Technically, this isn't our code because we just use the `backoff` crate for this. However, I'd rather have this tested here to make sure we are using it correctly.